js/modules/k6/experimental/streams/readable_streams.go

```diff
@@ -0,0 +1,586 @@
+func (stream *ReadableStream) setupReadableStreamDefaultControllerFromUnderlyingSource(
+	underlyingSource *goja.Object,
+	underlyingSourceDict UnderlyingSource,
+	highWaterMark float64,
+	sizeAlgorithm SizeAlgorithm,
+) {
+	// 1. Let controller be a new ReadableStreamDefaultController.
+	controller := &ReadableStreamDefaultController{}
+
+	// 2. Let startAlgorithm be an algorithm that returns undefined.
+	var startAlgorithm UnderlyingSourceStartCallback = func(*goja.Object) goja.Value {
+		return goja.Undefined()
+	}
+
+	// 3. Let pullAlgorithm be an algorithm that returns a promise resolved with undefined.
+	var pullAlgorithm UnderlyingSourcePullCallback = func(*goja.Object) *goja.Promise {
+		return newResolvedPromise(stream.vu, goja.Undefined())
+	}
+
+	// 4. Let cancelAlgorithm be an algorithm that returns a promise resolved with undefined.
+	var cancelAlgorithm UnderlyingSourceCancelCallback = func(any) goja.Value {
+		return stream.vu.Runtime().ToValue(newResolvedPromise(stream.vu, goja.Undefined()))
+	}
+
+	// 5. If underlyingSourceDict["start"] exists, then set startAlgorithm to an algorithm
+	// which returns the result of invoking underlyingSourceDict["start"] with argument
+	// list « controller » and callback this value underlyingSource.
+	if underlyingSourceDict.startSet {
+		startAlgorithm = stream.startAlgorithm(underlyingSource)
+	}
+
+	// 6. If underlyingSourceDict["pull"] exists, then set pullAlgorithm to an algorithm which
+	// returns the result of invoking underlyingSourceDict["pull"] with argument list
+	// « controller » and callback this value underlyingSource.
+	if underlyingSourceDict.pullSet {
+		pullAlgorithm = func(obj *goja.Object) (p *goja.Promise) {
+			if e := stream.runtime.Try(func() {
+				p = underlyingSourceDict.Pull(obj)
+			}); e != nil {
+				p = newRejectedPromise(stream.vu, e.Value())
+			}
+
+			if p == nil {
+				p = newResolvedPromise(stream.vu, goja.Undefined())
+			}
+			return p
+		}
+	}
+
+	// 7. If underlyingSourceDict["cancel"] exists, then set cancelAlgorithm to an algorithm which takes an argument
+	// reason and returns the result of invoking underlyingSourceDict["cancel"] with argument list « reason » and
+	// callback this value underlyingSource.
+	if underlyingSourceDict.cancelSet {
+		cancelAlgorithm = stream.cancelAlgorithm(underlyingSource)
+	}
+
+	// 8. Perform ? SetUpReadableStreamDefaultController(...)
+	stream.setupDefaultController(controller, startAlgorithm, pullAlgorithm, cancelAlgorithm, highWaterMark, sizeAlgorithm)
+}
+
+func (stream *ReadableStream) startAlgorithm(underlyingSource *goja.Object) UnderlyingSourceStartCallback {
+	call, ok := goja.AssertFunction(underlyingSource.Get("start"))
+	if !ok {
+		common.Throw(stream.runtime, errors.New("underlyingSource.[[start]] must be a function"))
+	}
+
+	return func(obj *goja.Object) (v goja.Value) {
+		var err error
+		v, err = call(underlyingSource, obj)
+		if err != nil {
+			panic(err)
```

Comment:
Out of curiosity, why is here (and in the following methods) a vanilla `panic`? :thinking:

Reply:
What `common.Throw` and our custom `throw` (see the description for further details) methods do is some manipulation over the error, and then `panic`. In these cases, we can just _throw_ whatever error we get, so we can just `panic` straightway.

Later, I think it's all the Goja wrapping who's responsible to adequately propagate that _throw_ to the JS surface.

Reply:
For additional context (correct me if I'm wrong joan), we `panic` directly here to avoid the wrapping involved in `common.Throw` because it keeps the WPT tests to pass, as they expect to find JS errors and receive GoError (goja type if I'm not mistaken) instead.